README
UTS
===

BTEC Unit Tracking System is a VBA based spreadsheet application used in the tracking of a class of pupils on a single unit. It provides a quick and painless way of setting up of a tracker, automatic grade based colouring and sorting of students as well as only displaying the exact number of criteria needed.


CHANGES
Changes:
v1.1
Changed dialogue boxes to have more accurate information.

v1.0.1
Added a course title textbox in the setup.
Added a unit title textbox in the setup.
Added a group title textbox in the setup.

v1.0
Initial release, fully working tracker though needs testing.


modules/Misc.bas
Attribute VB_Name = "Misc"
''  Program: Unit Tracking System (UTS)
''  Created by Simon Peter Campbell
''
''  The purpose of UTS is to provide an automated tracking system for individual BTEC units.
''  The aim is to create a fully functional and bug free system that is scalable in terms
''  of different amounts of criteria, students & sorting preferences.
''
''  All miscellaneous functions and procedures appear here. This module is for any algorithms
''  that don't fit into the other modules and are usually more general purpose.
''
''  Copyright 2012 Simon Peter Campbell
''  This file is part of Unit Tracking System (UTS).
''
''  Unit Tracking System is free software: you can redistribute it and/or modify it under
''  the terms of the GNU General Public License as published by the Free Software Foundation,
''  either version 3 of the License, or any later version. Unit Tracking System is distributed
''  in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
''  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
''  Public License for more details. You should have received a copy of the GNU General Public
''  License along with Unit Tracking System. If not, see <http://www.gnu.org/licenses/>.
''
''
''                       DD/MM/YYYY
''  Version 1.0.1 FINAL (27/09/2012)

Public Pass As Integer
Public Merit As Integer
Public Distinction As Integer
Public Students As Integer

Private Sub TrackerSettings_Click()
    Load frmSettings
    frmSettings.Show
End Sub

Function obtainColLetter(ByVal address As String) As String
'' A simple function that returns the column letter for any cell via its address function.
'' (columnabsolute:=false) must be specified or this won't work.
    obtainColLetter = Left(address & "", Application.WorksheetFunction.Find(":", address) - 1)
End Function

Sub writeVariables()
    variables.Select
    Dim Students As Range:      Set Students = Cells(6, 2)
    Dim Passes As Range:        Set Passes = Cells(7, 2)
    Dim Merits As Range:        Set Merits = Cells(8, 2)
    Dim Distinctions As Range:  Set Distinctions = Cells(9, 2)
    '' 10 is skipped because it is no longer needed
    Dim Assignments As Range:   Set Assignments = Cells(11, 2)
    Dim criteria As Range:      Set criteria = Cells(12, 2)
    Dim GradeColumn As Range:   Set GradeColumn = Cells(13, 2)
    Dim Bottom As Range:        Set Bottom = Cells(14, 2)
    Dim sortType As Range:      Set sortType = Cells(15, 2)
    Dim courseTitle As Range:   Set courseTitle = Cells(16, 2)
    Dim unitTitle As Range:     Set unitTitle = Cells(17, 2)
    Dim groupTitle As Range:    Set groupTitle = Cells(18, 2)
    
    Students.Value = frmSettings.numStudents.Value
    Passes.Value = frmSettings.numPass.Value
    Merits.Value = frmSettings.numMerit.Value
    Distinctions.Value = frmSettings.numDistinction.Value
    
    Dim numCriteria As Integer: numCriteria = Passes.Value + Merits.Value + Distinctions.Value
    Dim endCriteria As String: endCriteria = obtainColLetter(Cells(1, 4 + numCriteria).EntireColumn.address(columnabsolute:=False))
    
    Assignments.Value = "E7:" & endCriteria & 8
    criteria.Value = "E9:" & endCriteria & 8 + Students.Value
    GradeColumn.Value = obtainColLetter(Cells(1, 5 + numCriteria).EntireColumn.address(columnabsolute:=False))
    Bottom.Value = "B" & 9 + frmSettings.numStudents.Value
    If frmSettings.radAlphabet.Value = True Then sortType.Value = 1
    If frmSettings.radGrade.Value = True Then Cells("15", "B").Value = 2
    If frmSettings.radLeader.Value = True Then Cells("15", "B").Value = 3
    courseTitle.Value = frmSettings.txtCourse.Value
    unitTitle.Value = frmSettings.txtUnit.Value
    groupTitle.Value = frmSettings.txtGroup.Value
    Unit1.Select
End Sub

Sub determineCriteria(ByRef result As Integer, ByVal target As Range)
'' This procedure finds out what column the modified cell is in. This happens to make
'' sure that the cells are coloured correctly.
    Dim critLoop As Boolean: critLoop = True
    target.Select
    Do
        If Selection.Columns.count > 1 Then
            Select Case Selection.Cells(1, 1).Value
            Case "PASS"
                result = 1
            Case "MERIT"
                result = 2
            Case "DISTINCTION"
                result = 3
            End Select
        End If
        If Not ActiveCell.row = 1 Then
            Selection.Offset(-1, 0).Select
        Else
            critLoop = False
        End If
    Loop Until critLoop = False
End Sub

Sub StudentColour(ByVal row As Integer, ByVal target As Range)
'' A simple procedure that colours in an entire students row if they have achieved
'' a grade.
    Dim grade As Range: Set grade = Range(variables.Cells(13, 2).Value & row)
    Range("B" & row, "C" & row).Select
    Select Case grade.Value
    Case "Distinction"
        ColourDistinction
        Range(grade, grade.Offset(0, 2)).Select
        ColourDistinction '' colour the heading cells too.
    Case "Merit"
        ColourMerit
        Range(grade, grade.Offset(0, 2)).Select
        ColourMerit
    Case "Pass"
        ColourPass
        Range(grade, grade.Offset(0, 2)).Select
        ColourPass
    Case "Pass Referral"
        ColourPassR
        Range(grade, grade.Offset(0, 2)).Select
        ColourPassR
    Case "Unsafe"
        ColourRisk
        Range(grade, grade.Offset(0, 2)).Select
        ColourRisk
    Case Else
        If target.address <> ActiveCell.EntireRow.address Then
            ColourRemove
            Range(grade, grade.Offset(0, 2)).Select
            ColourRemove '' z is used for sorting reasons if no grade has been achieved
                         '' so the grade column is given white text to make it invisible.
            grade.Font.ThemeColor = xlThemeColorDark1
            grade.Font.TintAndShade = 0
        End If
    End Select
End Sub


modules/Macros.bas
Attribute VB_Name = "Macros"
''  Program: Unit Tracking System (UTS)
''  Created by Simon Peter Campbell
''
''  The purpose of UTS is to provide an automated tracking system for individual BTEC units.
''  The aim is to create a fully functional and bug free system that is scalable in terms
''  of different amounts of criteria, students & sorting preferences.
''
''  All recorded macros appear here, some have been modified to be modular in nature.
''
''  Copyright 2012 Simon Peter Campbell
''  This file is part of Unit Tracking System (UTS).
''
''  Unit Tracking System is free software: you can redistribute it and/or modify it under
''  the terms of the GNU General Public License as published by the Free Software Foundation,
''  either version 3 of the License, or any later version. Unit Tracking System is distributed
''  in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
''  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
''  Public License for more details. You should have received a copy of the GNU General Public
''  License along with Unit Tracking System. If not, see <http://www.gnu.org/licenses/>.
''
''
''                     DD/MM/YYYY
''  Version 1.0 FINAL (24/07/2012)

Sub addBorders(curRange As Range)
'
' Borders Macro
' Apply borders to cell
'
'
    curRange.Select
    Selection.Borders(xlDiagonalDown).LineStyle = xlNone
    Selection.Borders(xlDiagonalUp).LineStyle = xlNone
    With Selection.Borders(xlEdgeLeft)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeTop)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeBottom)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeRight)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlInsideVertical)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlInsideHorizontal)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
End Sub
Sub addBordersS()
'
' Borders Macro
' Apply borders to cell
'
'
    Selection.Borders(xlDiagonalDown).LineStyle = xlNone
    Selection.Borders(xlDiagonalUp).LineStyle = xlNone
    With Selection.Borders(xlEdgeLeft)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeTop)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeBottom)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlEdgeRight)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlInsideVertical)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
    With Selection.Borders(xlInsideHorizontal)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlThin
    End With
End Sub
Sub AddThickBorders()
'
' AddThickBorders Macro
'

'
    Selection.Borders(xlDiagonalDown).LineStyle = xlNone
    Selection.Borders(xlDiagonalUp).LineStyle = xlNone
    With Selection.Borders(xlEdgeLeft)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlMedium
    End With
    With Selection.Borders(xlEdgeTop)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlMedium
    End With
    With Selection.Borders(xlEdgeBottom)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlMedium
    End With
    With Selection.Borders(xlEdgeRight)
        .LineStyle = xlContinuous
        .ColorIndex = 0
        .TintAndShade = 0
        .Weight = xlMedium
    End With
    Selection.Borders(xlInsideVertical).LineStyle = xlNone
    Selection.Borders(xlInsideHorizontal).LineStyle = xlNone
End Sub

Sub ColourAss1()
'
' ColourAss1 Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .Color = 15773696
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourAss2()
'
' ColourAss2 Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent3
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourAss3()
'
' ColourAss3 Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .Color = 49407
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourAss4()
'
' ColourAss4 Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .Color = 65535
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourAss5()
'
' ColourAss5 Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .Color = 5142402
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourGrey()
'
' ColourGrey Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorDark1
        .TintAndShade = -0.249977111117893
        .PatternTintAndShade = 0
    End With
End Sub
Sub ColourDistinction()
'
' ColourDistinction Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent5
        .TintAndShade = -0.499984740745262
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorDark1
        .TintAndShade = 0
    End With
End Sub
Sub ColourDistinctionR()
'
' ColourDistinctionR Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent5
        .TintAndShade = 0.399975585192419
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorLight1
        .TintAndShade = 0
    End With
End Sub
Sub ColourMeritR()
'
' ColourMeritR Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent4
        .TintAndShade = 0.399975585192419
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorLight1
        .TintAndShade = 0
    End With
End Sub
Sub ColourPass()
'
' ColourPass Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent3
        .TintAndShade = -0.499984740745262
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorDark1
        .TintAndShade = 0
    End With
End Sub
Sub ColourPassR()
'
' ColourPassR Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .Color = 5296274
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorLight1
        .TintAndShade = 0
    End With
End Sub
Sub ColourRisk()
'
' ColourRisk Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent2
        .TintAndShade = -0.499984740745262
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorDark1
        .TintAndShade = 0
    End With
End Sub
Sub ColourMerit()
'
' ColourMerit Macro
'

'
    With Selection.Interior
        .Pattern = xlSolid
        .PatternColorIndex = xlAutomatic
        .ThemeColor = xlThemeColorAccent4
        .TintAndShade = -0.499984740745262
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorDark1
        .TintAndShade = 0
    End With
End Sub
Sub ColourRemove()
'
' ColourRemove Macro
'

'
    With Selection.Interior
        .Pattern = xlNone
        .TintAndShade = 0
        .PatternTintAndShade = 0
    End With
    With Selection.Font
        .ThemeColor = xlThemeColorLight1
        .TintAndShade = 0
    End With
End Sub
Sub SortAlphabet(ByVal SortRange As Range)
'
' SortAlphabet Macro
'

'
    
    ActiveSheet.Sort.SortFields.Clear
    ActiveSheet.Sort.SortFields.Add Key:=Range(Cells(SortRange.Cells(1).row, SortRange.Cells(1).Column).address, Cells(SortRange.Cells(SortRange.count).row, SortRange.Cells(1).Column).address) _
        , SortOn:=xlSortOnValues, Order:=xlAscending, DataOption:=xlSortNormal
    With ActiveSheet.Sort
        .SetRange SortRange
        .Header = xlGuess
        .MatchCase = True
        .Orientation = xlTopToBottom
        .SortMethod = xlPinYin
        .Apply
    End With
End Sub
Sub SortGrade(ByVal SortRange As Range)
'
' SortGrade Macro
'

'
    ActiveSheet.Sort.SortFields.Clear
    ActiveSheet.Sort.SortFields.Add Key:=Range(Cells(SortRange.Cells(1).row, SortRange.Cells(SortRange.count).Column).address, Cells(SortRange.Cells(SortRange.count).row, SortRange.Cells(SortRange.count).Column).address) _
        , SortOn:=xlSortOnValues, Order:=xlDescending, DataOption:=xlSortNormal
    ActiveSheet.Sort.SortFields.Add Key:=Range(Cells(SortRange.Cells(1).row, (SortRange.Cells(SortRange.count).Column - 2)).address, Cells(SortRange.Cells(SortRange.count).row, (SortRange.Cells(SortRange.count).Column - 2)).address) _
        , SortOn:=xlSortOnValues, Order:=xlAscending, DataOption:=xlSortNormal
    ActiveSheet.Sort.SortFields.Add Key:=Range(Cells(SortRange.Cells(1).row, SortRange.Cells(1).Column).address, Cells(SortRange.Cells(SortRange.count).row, SortRange.Cells(1).Column).address) _
        , SortOn:=xlSortOnValues, Order:=xlAscending, DataOption:=xlSortNormal
    With ActiveSheet.Sort
        .SetRange SortRange
        .Header = xlGuess
        .MatchCase = True
        .Orientation = xlTopToBottom
        .SortMethod = xlPinYin
        .Apply
    End With
End Sub
Sub ProtectSheet()
'
' ProtectSheet Macro
'

'
    ActiveSheet.Protect DrawingObjects:=True, Contents:=True, Scenarios:=True _
        , AllowFormattingCells:=True, AllowFormattingColumns:=True, _
        AllowFormattingRows:=True, Password:="21102560", UserInterfaceOnly:=True
End Sub





modules/Security.bas
Attribute VB_Name = "Security"
''  Program: Unit Tracking System (UTS)
''  Created by Simon Peter Campbell
''
''  The purpose of UTS is to provide an automated tracking system for individual BTEC units.
''  The aim is to create a fully functional and bug free system that is scalable in terms
''  of different amounts of criteria, students & sorting preferences.
''
''  This module covers everything to do with spreadsheet security. The locking of cells and
''  protecting of sheets is all done here.
''
''  Copyright 2012 Simon Peter Campbell
''  This file is part of Unit Tracking System (UTS).
''
''  Unit Tracking System is free software: you can redistribute it and/or modify it under
''  the terms of the GNU General Public License as published by the Free Software Foundation,
''  either version 3 of the License, or any later version. Unit Tracking System is distributed
''  in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
''  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
''  Public License for more details. You should have received a copy of the GNU General Public
''  License along with Unit Tracking System. If not, see <http://www.gnu.org/licenses/>.
''
''
''                    DD/MM/YYYY
''  Version 1.0 BETA (24/07/2012)
''
''  TODO:
''  1) Fully test the code - In Progress

Sub doSecurity()
    Dim Pass As Integer:          Pass = frmSettings.numPass.Value
    Dim Merit As Integer:         Merit = frmSettings.numMerit.Value
    Dim Distinction As Integer:   Distinction = frmSettings.numDistinction.Value
    Dim numCriteria As Integer:   numCriteria = Pass + Merit + Distinction
    Dim numStudents As Integer:   numStudents = frmSettings.numStudents.Value
    
    Dim topRow As Range:        Set topRow = Range("B6", Cells(6, 4 + numCriteria))
    Dim topTitles As Range:     Set topTitles = Range("B6", "D8")
    Dim criteriaRow As Range:   Set criteriaRow = Range("E8", Cells(8, 4 + numCriteria))
    Dim support As Range:       Set support = Range("D9", Cells(8 + numStudents, 4))
    Dim gradeCol As Range:      Set gradeCol = Range(Cells(6, 5 + numCriteria), Cells(8 + numStudents, 5 + numCriteria))
    Dim notesCell As Range:     Set notesCell = Range(Cells(6, 5 + numCriteria + 1), Cells(8, 5 + numCriteria + 1))
    Dim pointCol As Range:      Set pointCol = Range(Cells(6, 5 + numCriteria + 2), Cells(8 + numStudents, 5 + numCriteria + 2))
    Dim courseTitle As Range:   Set courseTitle = Cells(2, 5 + numCriteria)
    Dim unitGroup As Range:     Set unitGroup = Range(Cells(4, 5 + numCriteria), Cells(4, 5 + numCriteria + 1))
    
    Unit1.Cells.Locked = False
    variables.Cells.Locked = False
    topRow.Locked = True
    topTitles.Locked = True
    criteriaRow.Locked = True
    support.Locked = True
    gradeCol.Locked = True
    notesCell.Locked = True
    pointCol.Locked = True
    courseTitle.MergeArea.Locked = True
    unitGroup.Locked = True
        
    Unit1.Activate: ProtectSheet '' This is a recorded macro
    variables.Activate: ProtectSheet
    variables.Visible = xlSheetHidden
    Unit1.Activate
End Sub

Sub unlockSheets()
    Unit1.Unprotect Password:="21102560"
    variables.Unprotect Password:="21102560"
End Sub



modules/Extras.bas
Attribute VB_Name = "Extras"
''  Program: Unit Tracking System (UTS)
''  Created by Simon Peter Campbell
''
''  The purpose of UTS is to provide an automated tracking system for individual BTEC units.
''  The aim is to create a fully functional and bug free system that is scalable in terms
''  of different amounts of criteria, students & sorting preferences.
''
''  This module contains the code that formats all the additional headings such as "overall
''  grade". It also fills in relevant formulas and formats the points column.
''
''  Copyright 2012 Simon Peter Campbell
''  This file is part of Unit Tracking System (UTS).
''
''  Unit Tracking System is free software: you can redistribute it and/or modify it under
''  the terms of the GNU General Public License as published by the Free Software Foundation,
''  either version 3 of the License, or any later version. Unit Tracking System is distributed
''  in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
''  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
''  Public License for more details. You should have received a copy of the GNU General Public
''  License along with Unit Tracking System. If not, see <http://www.gnu.org/licenses/>.
''
''
''                       DD/MM/YYYY
''  Version 1.0.1 FINAL (27/09/2012)
''
''  TODO:
''

Sub DoExtras()
    Dim Pass As Integer:          Pass = frmSettings.numPass.Value
    Dim Merit As Integer:         Merit = frmSettings.numMerit.Value
    Dim Distinction As Integer:   Distinction = frmSettings.numDistinction.Value
    Dim numCriteria As Integer:   numCriteria = Pass + Merit + Distinction
    Dim gradeCell As Range:   Set gradeCell = Cells(6, (5 + numCriteria))
    Dim notesCell As Range:   Set notesCell = gradeCell.Offset(0, 1)
    Dim pointCell As Range:   Set pointCell = notesCell.Offset(0, 1)
    Dim unitsCell As Range:   Set unitsCell = gradeCell.Offset(-4, 0)
    Dim keyHome As Range:     Set keyHome = unitsCell.Offset(0, 3)
    
    formatGrade gradeCell
    formatNotes notesCell
    formatPoint pointCell
    formatUnitTitles unitsCell
    formatExtras frmSettings.numStudents.Value
    formatKey keyHome
End Sub

Private Sub formatGrade(ByVal heading As Range)
'' Format the grades heading and assign the correct formula
    Dim Pass As Integer:          Pass = frmSettings.numPass.Value
    Dim Merit As Integer:         Merit = frmSettings.numMerit.Value
    Dim Distinction As Integer:   Distinction = frmSettings.numDistinction.Value
    Dim numCriteria As Integer:   numCriteria = Pass + Merit + Distinction
    Dim numStudents As Integer:   numStudents = frmSettings.numStudents.Value
    Dim PassEnd As String:        PassEnd = obtainColLetter(Cells(1, 4 + Pass).EntireColumn.address(columnabsolute:=False))
    Dim MeritEnd As String:       MeritEnd = obtainColLetter(Cells(1, 4 + Pass + Merit).EntireColumn.address(columnabsolute:=False))
    Dim DistinctionEnd As String: DistinctionEnd = obtainColLetter(Cells(1, 4 + numCriteria).EntireColumn.address(columnabsolute:=False))
    
    heading.UnMerge
    Range(heading.address, heading.Offset(2, 0)).Select
    ColourGrey
    With Selection
        .Merge
        .Value = "Overall Grade"
        .Font.Bold = True
        .Font.Size = 16
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .EntireColumn.ColumnWidth = 25
    End With
    Range(heading.Offset(1, 0), heading.Offset(numStudents, 0)).Select
    With Selection
        .ClearFormats
        .Font.Bold = True
        .Font.ThemeColor = xlThemeColorDark1
        .Font.TintAndShade = 0
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
    End With
    
    Dim r As Integer: r = 9 '' The start of the student rows.
    Dim curCell As Range
    For Each curCell In Selection
'       Now the incredibly complex formula for calculating grades! I doubt it is understandable
'       so I broke it down in comment form with a mixture of Excel and pseudocode. It shows the
'       steps taken in determining the grade.
'       curCell.Formula = "=
'       IF(COUNTIF(criteria range = 'R')= number-criteria, "Distinction" grade,
'       IF(COUNTIF(pass + merit range = 'R')= pass+merit critera, "Merit" grade,
'       IF(COUNTIF(pass range = 'R')= pass criteria, "Pass" grade,
'       IF(SUM(COUNTIF(pass range = 'R'), COUNTIF(pass range = '8'))=pass criteria, "Pass Referral", _
'       IF(SUM(COUNTIF(pass range = 'R'), COUNTIF(pass range = 'T'), COUNTIF(pass range = '8'))=pass criteria, "Unsafe", "z" for sorting purposes))))))"
        curCell.Formula = "=IF(COUNTIF($E" & r & ":$" & DistinctionEnd & r & ", ""R"")=" & numCriteria & ", ""Distinction"",IF(COUNTIF($E" & r & ":$" & MeritEnd & r & ", ""R"")=" & Pass + Merit & ", ""Merit"",IF(COUNTIF($E" & r & ":$" & PassEnd & r & ", ""R"")=" & Pass & ", ""Pass"", IF(SUM(COUNTIF($E" & r & ":$" & PassEnd & r & ", ""R""), COUNTIF($E" & r & ":$" & PassEnd & r & ", ""8""))=" & Pass & ", ""Pass Referral"", IF(SUM(COUNTIF($E" & r & ":$" & PassEnd & r & ", ""R""), COUNTIF($E" & r & ":$" & PassEnd & r & ", ""T""), COUNTIF($E" & r & ":$" & PassEnd & r & ", ""8""))=" & Pass & ", ""Unsafe"", ""z"")))))"
        StudentColour r, Cells(curCell.row, curCell.Column)
        r = r + 1
    Next curCell
    addBorders Range(heading.address, heading.Offset(numStudents, 0))
End Sub

Private Sub formatNotes(ByVal heading As Range)
'' This procedure formats the notes column.
    Dim numStudents As Integer: numStudents = frmSettings.numStudents.Value
    heading.UnMerge
    Range(heading, heading.Offset(2, 0)).Select
    ColourGrey
    With Selection
        .Merge
        .Value = "Notes"
        .Font.Bold = True
        .Font.Size = 16
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .EntireColumn.ColumnWidth = 25
    End With
    With Range(heading.Offset(1, 0), heading.Offset(numStudents, 0))
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
    End With
    addBorders Range(heading, heading.Offset(numStudents, 0))
End Sub

Private Sub formatPoint(ByVal heading As Range)
'' Simple procedure to give the correct formatting and formula to every cell in
'' the points heading.
    Dim Pass As Integer:          Pass = frmSettings.numPass.Value
    Dim Merit As Integer:         Merit = frmSettings.numMerit.Value
    Dim Distinction As Integer:   Distinction = frmSettings.numDistinction.Value
    Dim numCriteria As Integer:   numCriteria = Pass + Merit + Distinction
    Dim numStudents As Integer:   numStudents = frmSettings.numStudents.Value
    Dim PassEnd As String:        PassEnd = obtainColLetter(Cells(1, 4 + Pass).EntireColumn.address(columnabsolute:=False))
    Dim MeritEnd As String:       MeritEnd = obtainColLetter(Cells(1, 4 + Pass + Merit).EntireColumn.address(columnabsolute:=False))
    Dim DistinctionEnd As String: DistinctionEnd = obtainColLetter(Cells(1, 4 + numCriteria).EntireColumn.address(columnabsolute:=False))
    Dim sortType As Integer:      sortType = variables.Cells(15, 2).Value
    Dim gradeLetter As String:    gradeLetter = obtainColLetter(heading.Offset(0, -2).EntireColumn.address(columnabsolute:=False))
    Dim r As Integer: r = 9 '' The first student row
    '' Format all of the cells and add borders to them.
    heading.UnMerge
    Range(heading, heading.Offset(2, 0)).Select
    ColourGrey
    With Selection
        .Merge
        .Value = "Points"
        .Font.Bold = True
        .Font.Size = 12
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlLeft
        .Orientation = -90
    End With
    Range(heading.Offset(1, 0), heading.Offset(numStudents, 0)).Select
    With Selection
        .Font.Bold = True
        .Font.Size = 11
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .EntireColumn.ColumnWidth = 2.29
    End With
    '' Next add the correct formula depending on the users choice.
    Dim curCell As Range
    For Each curCell In Selection
        Select Case sortType
        Case 1 '' 1 is alphabetic
            curCell.Formula = 1
        Case 2 '' 2 is grade hence the complicated formula
            curCell.Formula = "=IF($" & gradeLetter & r & " = ""Distinction"", 4, IF($" & gradeLetter & r & " = ""Merit"", 3, IF($" & gradeLetter & r & " = ""Pass"",2,IF($" & gradeLetter & r & " = ""Pass Referral"",1,IF($" & gradeLetter & r & " = ""z"",0,IF($" & gradeLetter & r & " = ""Unsafe"",-1,0))))))"
        Case 3 '' 3 is points, 3 points for a pass, 1 for referral and -1 for a missed deadline
            curCell.Formula = "=SUM(COUNTIF($E" & r & ":$" & DistinctionEnd & r & ", ""R"") * 3, COUNTIF($E" & r & ":$" & DistinctionEnd & r & ", ""8"") * 1, COUNTIF($E" & r & ":$" & DistinctionEnd & r & ", ""T"") * -1)"
        End Select
        r = r + 1
    Next curCell
    '' Since the points column isn't necessary for grade of alphabetic, hide the column.
    If sortType = 1 Or sortType = 2 Then
        ActiveCell.EntireColumn.Hidden = True
    ElseIf sortType = 3 Then
        ActiveCell.EntireColumn.Hidden = False
    End If
    addBorders Range(heading.Offset(0, -1), heading.Offset(numStudents, 0))
End Sub

Private Sub formatUnitTitles(ByVal heading As Range)
'' The extra titles that display the unit name, group code and course title are formatted here.
    Dim workingRange As Range: Set workingRange = Range(heading, heading.Offset(3, 0))
    workingRange.UnMerge
    With Range(workingRange.Cells(1, 1), workingRange.Cells(1, 2))
        .Merge
        .Value = "Course Title"
        .Font.Bold = True
        .Font.Size = 12
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .Select
    End With
    ColourGrey
    With Range(workingRange.Cells(2, 1), workingRange.Cells(2, 2))
        .Merge
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .Value = frmSettings.txtCourse.Value
    End With
    With Range(workingRange.Cells(3, 1), workingRange.Cells(3, 2))
        .Font.Bold = True
        .Font.Size = 12
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .Select
    End With
    ColourGrey
    workingRange.Cells(3, 1).Value = "Unit"
    workingRange.Cells(3, 2).Value = "Group"
    With Range(workingRange.Cells(4, 1), workingRange.Cells(4, 2))
        .Font.Size = 12
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .ShrinkToFit = True
    End With
    workingRange.Cells(4, 1).Value = frmSettings.txtUnit.Value
    workingRange.Cells(4, 2).Value = frmSettings.txtGroup.Value
    addBorders workingRange
End Sub

Private Sub formatExtras(ByVal Students As Integer)
    With Range(Cells(4, 2), Cells(5, 3))
        .Merge
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .Value = "Unit Tracking System"
    End With
    
    Dim workingRange As Range: Set workingRange = Range(Cells(10 + Students, 2), Cells(10 + Students + 2, 3))
    With workingRange
        .Merge
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
        .Select
    End With
    If workingRange.Cells(1, 1).Value = "" Then workingRange.Value = "ENTER MESSAGE HERE!"
    AddThickBorders
End Sub

Private Sub formatKey(ByVal home As Range)
    With Range(home, home.Offset(2, 2))
        .Clear
        .ColumnWidth = 2.71
    End With
    With Range(home, home.Offset(2, 0))
        .VerticalAlignment = xlCenter
        .HorizontalAlignment = xlCenter
    End With
    
    Dim sortType As Integer: sortType = variables.Cells(15, 2).Value
    If sortType > 3 Then Set home = home.Offset(0, 1)
    
    Dim workingRange As Range: Set workingRange = Range(home, home.Offset(2, 1))
    Range(home, home.Offset(2, 0)).Font.name = "Wingdings 2"
    workingRange.Cells(1, 1).Value = "R"
    workingRange.Cells(2, 1).Value = "8"
    workingRange.Cells(3, 1).Value = "T"
    workingRange.Cells(1, 2).Value = "Pass (R)"
    workingRange.Cells(2, 2).Value = "Referral (8)"
    workingRange.Cells(3, 2).Value = "Deadline Missed (T)"
    workingRange.Cells.Locked = True
End Sub
